### DISCUSSION TITLE
Using `.` as separator in property names (after prefix)

### DISCUSSION THREAD

**Discussion by ppatierno** [quarkusio/quarkus#31156] [Q&A] [answered]
Hello,
I know that when using @ConfigMapping we are relying on Smallrye but I am having troubles with naming keys in a Map field.
From the doc here https://quarkus.io/guides/config-mappings#maps I would like to have the names of my properties using `.` as separator.
Referring to the example on the doc, is it possible to have something like this?

```shell
server.form.landing.page=index.html
```

So I would like to have all stuff coming after `server.form` (the config prefix), to be used as key in my map, so the key would be `landing.page` (and not `landing-page` as in the example).

I am already using `smallrye.config.mapping.validate-unknown=false` to avoid validation errors, but anyway my Map is empty.

I know it's possible by putting double quotes around but it doesn't work for us because we are porting to quarkus something we already have in place and cannot ask to users to change all their parameters.

**Comment by radcortez** [comment]
We don't support that at the moment, but I guess that if you have a mapping type of `Map<String, String>` (or any other Leaf element), we could assume that everything after the mapping path belongs to the key. For nested elements, we cannot do such an assumption, because we don't have a way to know when the key ends and the nested properties elements start.

**Reply by ppatierno** [reply]
@radcortez I actually have something like this ..

```java
@ConfigMapping(prefix = "kafka")
public interface KafkaConfig {

    Map<String, String> admin();
    Map<String, String> producer();
    Map<String, String> consumer();
}
```

in order to have into the Map all the configurations with `kafka.consumer.` prefix but it's not working, unless I use the double quotes around the specific property (i.e. `kafka.consumer."auto.offset.reset"` instead of just `kafka.consumer.auto.offset.reset`).
If you think it should work instead, maybe there is something I am doing wrong on my side?

**Reply by radcortez** [reply]
No, that does not work without the quotes currently.

I'm thinking that we can maybe support it for this particular case because for a `Map<String, String>`, there is nothing else we need from the key, so we could consider the rest of the name to be the key.

Alternatively, you could go with the interceptor approach and quote the keys.

Anyway, let me have a look and see if we can support it without breaking something.

**Reply by ppatierno** [reply]
> there is nothing else we need from the key, so we could consider the rest of the name to be the key

Exactly, it would make more sense. Anyway replacing they key with using quotes with interceptor seems to work.
The drawback is the need for the `smallrye.config.mapping.validate-unknown=false` to have it working as well.

**Reply by ppatierno** [reply]
Adding one more thing ... (sorry for all these questions but doing this "porting" to Quarkus is raising more issues one after the other :-)) ...
Our current config have something like this:

```shell
#Apache Kafka common
kafka.bootstrap.servers=localhost:9092

#Apache Kafka producer
kafka.producer.acks=1

#Apache Kafka consumer
kafka.consumer.auto.offset.reset=earliest
```

so what is prefixed with `kafka.consumer` and `kafka.producer` which is what we were talking about.
I forgot we have common configuration (applied to both consumer and producer) which has `kafka.` prefix.
It means that for example `bootstrap.servers` should be treated as a key as well.
But the `KafkaConfig` class I described before there is no way to put all the `kafka.<properties>` into a corresponding Map. We are able to map `kafka.consumer.<propertie>` in a Map (as well as the producer) but I am not able to find a way to say "everything after `kafka.` prefix go into this map".
To summarinze:

* map `KafkaConfig.admin()` --> gets all properties under `kafka.admin`
* map `KafkaConfig.consumer()` --> gets all properties under `kafka.consumer`
* map `KafkaConfig.producer()` --> gets all properties under `kafka.producer`
* which map in `KafkaConfig` could get all properties under `kafka.` ??

**Reply by ppatierno** [reply]
@radcortez the only way I found was having 4 different classes like this :-(

```java
@ConfigMapping
public interface KafkaConfig {
    Map<String, String> kafka();
}

@ConfigMapping(prefix = "kafka")
public interface KafkaConsumerConfig {
    Map<String, String> consumer();
}

@ConfigMapping(prefix = "kafka")
public interface KafkaProducerConfig {
    Map<String, String> producer();
}

@ConfigMapping(prefix = "kafka")
public interface KafkaAdminConfig {
    Map<String, String> admin();
}
```

which means 4 injection points where they are needed

```java
@Inject
KafkaConfig kafkaConfig;

@Inject
KafkaConsumerConfig kafkaConsumerConfig;

@Inject
KafkaProducerConfig kafkaProducerConfig;

@Inject
KafkaAdminConfig kafkaAdminConfig;
```

Is there no way to group them all somehow?

**Reply by radcortez** [reply]
> Adding one more thing ... (sorry for all these questions but doing this "porting" to Quarkus is raising more issues one after the other :-)) ... Our current config have something like this:
 
No worries, I'm here to help :)

Are you able to get me a full typical configuration example in properties format? It would probably be easier for me to look at it and figure out the best way to map it.

> Is there no way to group them all somehow?

Yes, this should work:

```java
    @Test
    void kafkaConfig() {
        SmallRyeConfig config = new SmallRyeConfigBuilder()
            .addDefaultInterceptors()
            .withMapping(KafkaConfig.class)
            .withSources(config(
                "kafka.\"bootstrap.servers\"", "localhost:9092",
                "kafka.producer.\"acks\"", "1",
                "kafka.consumer.\"auto.offset.reset\"", "earliest"))
            .build();

        KafkaConfig mapping = config.getConfigMapping(KafkaConfig.class);

        assertEquals("localhost:9092", mapping.common().get("bootstrap.servers"));
        assertEquals("1", mapping.producer().get("acks"));
        assertEquals("earliest", mapping.consumer().get("auto.offset.reset"));
    }

    @ConfigMapping(prefix = "kafka")
    interface KafkaConfig {
        @WithParentName
        Map<String, String> common();
        Map<String, String> admin();
        Map<String, String> producer();
        Map<String, String> consumer();
    }
```

**Reply by radcortez** [reply]
The recommended alternative:

```java
    @ConfigMapping(prefix = "kafka")
    interface KafkaConfig {
        Bootstrap bootstrap();

        Producer producer();

        Consumer consumer();

        interface Bootstrap {
            List<String> servers();
        }

        interface Producer {
            Integer acks();
        }

        interface Consumer {
            @WithName("auto.offset.reset")
            String autoOffsetReset();
        }
    }
```

But if I remember correctly, Kafka expected the configuration in mapping format?

**Reply by ppatierno** [reply]
The first approach grouping everything in one class works like a charm! I didn't get the usage of the `@WithParentName` annotation!
The recommended one doesn't scale well, because we want to support all the Kafka config parameters and it's not feasible having a field for each of them, so using a Map is the best way (which is actually what I have in my own "no Quarkus" implementation).
And yes you are right, the Kafka clients gets Properties map for configuration.
So it seems I am close to have a working approach :-) Thanks!

**Reply by radcortez** [reply]
Great. So we need the quoted keys to be more transparent. I was not able to look into it yet. I'll let you know.

**Reply by ppatierno** [reply]
I am using an interceptor for quoting as you suggested ....

```java
public class BridgeConfigSourceInterceptor implements ConfigSourceInterceptor {

    @Override
    public ConfigValue getValue(ConfigSourceInterceptorContext context, String s) {
        ConfigValue cv = context.proceed(s);
        if (s.startsWith(KafkaConsumerConfig.KAFKA_CONSUMER_CONFIG_PREFIX)) {
            return cv.withName(this.withQuotes(KafkaConsumerConfig.KAFKA_CONSUMER_CONFIG_PREFIX, s));
        }
        if (s.startsWith(KafkaProducerConfig.KAFKA_PRODUCER_CONFIG_PREFIX)) {
            return cv.withName(this.withQuotes(KafkaProducerConfig.KAFKA_PRODUCER_CONFIG_PREFIX, s));
        }
        if (s.startsWith(KafkaAdminConfig.KAFKA_ADMIN_CONFIG_PREFIX)) {
            return cv.withName(this.withQuotes(KafkaAdminConfig.KAFKA_ADMIN_CONFIG_PREFIX, s));
        }
        if (s.startsWith(KafkaConfig.KAFKA_CONFIG_PREFIX)) {
            return cv.withName(this.withQuotes(KafkaConfig.KAFKA_CONFIG_PREFIX, s));
        }
        if (s.equals(HttpConfig.HTTP_PORT)) {
            return cv.withName("quarkus.http.port");
        }
        if (s.equals(HttpConfig.HTTP_HOST)) {
            return cv.withName("quarkus.http.host");
        }
        return cv;
    }

    private String withQuotes(String prefix, String key) {
        return String.format("%s\"%s\"", prefix, key.substring(prefix.length()));
    }
}
```

I am also mapping our `http.port` to the `quarkus.http.port` so that the user can still use our parameter instead of the quarkus one to configure the HTTP server.

**Reply by ppatierno** [reply]
@radcortez sorry I just noticed your comment on the GitHub issue about extending the relocate and fallback interceptors. I will look into them instead of having the one above :-)

**Reply by radcortez** [reply]
Yes, the quotes rewriting is OK. My recommendation was for the `http.port` and maybe others, which you could use the relocate or the fallback interceptor directly.

**Reply by ppatierno** [reply]
@radcortez I just created a relocate interceptor like this:

```java
public class BridgeConfigRelocateInterceptor extends RelocateConfigSourceInterceptor {
    public BridgeConfigRelocateInterceptor(Map<String, String> mappings) {
        // code ....
    }
}
```

But the application fails with:

```shell
(main) Failed to start quarkus: java.util.ServiceConfigurationError: io.smallrye.config.ConfigSourceInterceptor: io.strimzi.kafka.bridge.quarkus.config.BridgeConfigRelocateInterceptor Unable to get public no-arg constructor
        at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:582)
        at java.base/java.util.ServiceLoader.getConstructor(ServiceLoader.java:673)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1233)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1265)
        at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1300)
        at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1385)
        at io.smallrye.config.SmallRyeConfigBuilder.discoverInterceptors(SmallRyeConfigBuilder.java:134)
        at io.smallrye.config.SmallRyeConfig$ConfigSources.buildInterceptors(SmallRyeConfig.java:599)
        at io.smallrye.config.SmallRyeConfig$ConfigSources.<init>(SmallRyeConfig.java:545)
        at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:68)
        at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:532)
        at io.quarkus.runtime.configuration.QuarkusConfigFactory.getConfigFor(QuarkusConfigFactory.java:36)
        at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:76)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:181)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:331)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:252)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:60)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:86)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:447)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:59)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:104)
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:131)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
Caused by: java.lang.NoSuchMethodException: io.strimzi.kafka.bridge.quarkus.config.BridgeConfigRelocateInterceptor.<init>()
        at java.base/java.lang.Class.getConstructor0(Class.java:3349)
        at java.base/java.lang.Class.getConstructor(Class.java:2151)
        at java.base/java.util.ServiceLoader$1.run(ServiceLoader.java:660)
        at java.base/java.util.ServiceLoader$1.run(ServiceLoader.java:657)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/java.util.ServiceLoader.getConstructor(ServiceLoader.java:668)
        ... 22 more
```

**Reply by radcortez** [reply]
You need a public no-args constructor if the interceptor is registered via the ServiceLoader. If you want to bootstrap it yourself, please use `ConfigSourceInterceptorFactory`. This will give you access to the current `Config` when building the interceptor. Useful to provide configuration.

**Comment by radcortez** [comment, marked as answer]
With https://github.com/smallrye/smallrye-config/pull/900, SmallRye Config will support dotted keys without quotes for non-nested Map elements.

**Reply by ppatierno** [reply]
That's cool! Great work @radcortez !
